src/Compiler/CodeGen/IlxGen.fs

```diff
@@ -610,16 +610,13 @@ type PtrsOK =
     | PtrTypesNotOK
 
 let GenReadOnlyAttribute (g: TcGlobals) =
-    mkILCustomAttribute (g.attrib_IsReadOnlyAttribute.TypeRef, [], [], [])
+    if g.attrib_IsReadOnlyAttribute.TyconRef.CanDeref then
+        mkILCustomAttribute (g.attrib_IsReadOnlyAttribute.TypeRef, [], [], [])
+    else
+        mkILCustomAttribute (g.local_ReadOnlyAttribute_tcref(), [], [], [])
 
 let GenReadOnlyAttributeIfNecessary (g: TcGlobals) ty =
-    let add =
-        false
-        && g.isSystem_Runtime_CompilerServices_IsReadOnlyAttributeAvailable
-        && isInByrefTy g ty
-        && g.attrib_IsReadOnlyAttribute.TyconRef.CanDeref
-
-    if add then
+    if isInByrefTy g ty then
         let attr = GenReadOnlyAttribute g
```

Comment:
Will this reuse local one once generated for the first time?

Reply:
Generation is lazy and singleton, so yes.

Reply:
A point of clarification, GenReadOnlyAttribute  adds a reference to the IsReadOnlyAttribute to a type, which may occur many times in a compilation.  The referenced IsReadOnlyAttributed, is generated on demand in tcGlobals, it is generated only when a public IsReadOnlyAttribute is not found in any of the referenced assemblies, it is generated once per compilation unit (I.e. assembly).